crates/codegen/syntax_templates/src/typescript/cursor_ts_wrappers.rs

```diff
@@ -0,0 +1,160 @@
+use super::{
+    cst_ts_wrappers::{TextIndex, TextRange, ToJS},
+    cursor::Cursor as RustCursor,
+    kinds::*,
+};
+
+use napi::bindgen_prelude::*;
+use napi::JsObject;
+
+#[napi]
+pub struct Cursor(Box<RustCursor>);
+
+#[napi]
+impl Cursor {
+    pub(crate) fn new(cursor: RustCursor) -> Self {
+        Self(Box::new(cursor))
+    }
+
+    #[napi]
+    pub fn spawn(&self) -> Self {
+        Self::new(self.0.spawn())
+    }
+
+    #[napi(getter)]
+    pub fn is_finished(&self) -> bool {
+        self.0.is_finished()
+    }
+
+    #[napi(getter, ts_return_type = "RuleNode | TokenNode | null")]
+    pub fn node(&self, env: Env) -> Option<JsObject> {
```

Comment:
I think I'm missing something. This would convert it to a plain JS object, however, us typing it with `RuleNode | TokenNode` means that it will have the usual APIs on it, like the ones that need round tripping to Rust. How does that work?

this is related to other `to_js()` uses as well.

Reply:
In this case we have to return `JsObject` because the result is polymorphic. And fundamentally, `JsObject` is what is passed even if the Rust type were explicit.

Reply:
So on the receiver side, working in NodeJS land, they would get a plain JS object.
It wouldn't have the native Rust APIs? not sure how would calling its methods would round trip.

... I think I will wait for this to be ready and play with it for a bit to understand more.